### PR TITLE
fix(template): migrate open_source from legacy visibility answers on update

### PR DESCRIPTION
**Pull request by ichoosetoaccept**
## Summary

- Adds a `_migrations` step (`migrations/0.41.0_open_source.py`) that derives the `open_source` answer from the retired `project_visibility`/`project_audience` questions during `copier update`, instead of silently falling back to `open_source`'s schema default.
- Corrects a `copier.yml` comment that overstated `copier update`'s merge guarantees for `pyproject.toml`/`prek.toml`, and documents the real risk plus a mitigation in `docs/update.md`, along with a "never hand-edit `.copier-answers.yml`" note.
- Updates dependencies to latest (notably `copier` 9.15.0 → 9.16.0, `ruff`, `ty`), and fixes a pre-existing test-code type error the `ty` bump surfaced.

## Why

Without the migration, `copier update --defaults` resolved the new `open_source` question (added when `project_audience`/`project_visibility`/`use_community_health_files` were folded together) from its own schema default (`repository_provider == 'github.com'`) rather than the project's real history. Reproduced end-to-end: an **internal** GitHub project silently gained a LICENSE and all community-health files on update, with zero `.rej` files to flag it.

Investigated the deeper "silent overwrite" class of bug this surfaced (large template diffs overwriting hand-customized `pyproject.toml`/`prek.toml` sections) by reading Copier's own update algorithm (`git apply --reject` with a `git merge-file` fallback only for files that already produced a `.rej`) and checking upstream (`copier-org/copier#2376`, a stale draft PR that would replace this with a real `git merge`-based 3-way merge). No adoptable upstream fix exists today; tracked as [DOT-599](https://linear.app/ismar/issue/DOT-599/copier-update-can-silently-overwrite-hand-customized) for the remaining structural-merge and `context_lines` mitigation work.

## Test plan

- [x] `poe check` (lint + typecheck) passes
- [x] `poe test` (147 fast tests) and `poe test-all` (+5 slow integration tests) pass
- [x] `prek run --all-files` passes across the whole repo
- [x] Migration verified end-to-end against a temporary local `0.41.0` git tag in both directions: an internal-on-GitHub project stays internal, a public-on-GitLab project stays public
- [x] 6 new regression tests (`TestOpenSourceMigration`) unit-test the migration script directly, since `_migrations` are gated on the *tagged* template version and can't be exercised via a real `copier update` against an unreleased commit

Refs DOT-599

> [!NOTE]
> ### Migrate `open_source` answer from legacy visibility keys in `.copier-answers.yml` on update
> - Adds [migrations/0.41.0_open_source.py](https://github.com/detailobsessed/copier-uv-bleeding/pull/328/files#diff-9d59468c7b79ba209f1ade6a1c10145fedc830ad2dc1c06070704e0d5ac3100e), a migration script that derives `open_source: true|false` from legacy `project_visibility` or `project_audience` keys and appends it to `.copier-answers.yml`.
> - Registers the migration in [copier.yml](https://github.com/detailobsessed/copier-uv-bleeding/pull/328/files#diff-a74d725f30a732f10530aec0b66cdc1f1856272e35ede5ba4826ead3075c3d3d) to run before the copy stage when updating across version 0.41.0.
> - The script is a no-op if `open_source` already exists, the answers file is missing, no legacy keys are present, or legacy keys conflict with each other.
> - Adds tests in [tests/test_template.py](https://github.com/detailobsessed/copier-uv-bleeding/pull/328/files#diff-140d5ca36c77559547b5932938faad8b1296414593f6f0da782ab154c9637ea9) covering all derivation paths and edge cases.
> - Appends update guidance to [docs/update.md](https://github.com/detailobsessed/copier-uv-bleeding/pull/328/files#diff-05ad2f061c8db8d1e79bff07c7103fd0ebc273efb5fe7443c3fd9a42e818cad3) about silent overwrites in `pyproject.toml`/`prek.toml` and warns against hand-editing `.copier-answers.yml`.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 73af60a.</sup>
> 
>

### CODE REVIEW

File: copier.yml [hunk 1]
```diff
@@ -48,6 +54,18 @@ _exclude:
...
 
+# Migrations run once, during `copier update`, when crossing the given version
+# boundary. `open_source` (added in 0.41.0) replaced project_audience /
+# project_visibility / use_community_health_files. Without this, `copier
+# update --defaults` would resolve `open_source` from its own schema default
+# instead of the project's real history — silently flipping visibility for
+# any project whose history disagrees with that default. See
+# migrations/0.41.0_open_source.py.
+_migrations:
+- version: 0.41.0
```
[the remark this answers is not in the document]
Comment (ichoosetoaccept) [reply copier.yml:65]: Not applicable, for two reasons:

1. This concern is stale relative to the current push. `pyproject.toml` on this branch's head already reads `version = "0.41.0"` (this repo's own `chore(release): 0.41.0` landed on `main` back on 2026-06-02, when the original commit in this PR — `d074c60` — was squash-merged as #327; I rebased this branch onto that before the last push).
2. More fundamentally, Copier's `_migrations` version gating (`version: 0.41.0` in `copier.yml`) has nothing to do with this repo's own `pyproject.toml`. It's evaluated against the *template's git tags* — the version a downstream project is updating from/to, resolved via `copier update`'s VCS ref resolution — completely independent of this repository's own packaging metadata. I verified this directly: I created a temporary local git tag `0.41.0` on this branch and ran a real `copier copy`/`copier update` cycle against it (in both directions — an internal-on-GitHub project and a public-on-GitLab project), confirming the migration fires correctly purely based on that tag, with no dependency on `pyproject.toml`'s `version` field at all.

No change needed here.

File: migrations/0.41.0_open_source.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+ANSWERS_FILE = Path(".copier-answers.yml")
+
+
+def _extract(text: str, key: str) -> str | None:
+    match = re.search(rf"^{re.escape(key)}:\s*(\S+)\s*$", text, re.MULTILINE)
+    return match.group(1).strip("'\"") if match else None
+
+
+def main() -> None:
```
[the remark this answers is not in the document]
Comment (ichoosetoaccept) [reply migrations/0.41.0_open_source.py:40]: Good catch — you're right that this was a real gap. Fixed in 73af60a.

`use_community_health_files` is now incorporated: it's used as a fallback signal when `project_visibility`/`project_audience` are both absent (pre-#322 projects), and as a tie-breaker check when a visibility-derived value *is* available. When the two disagree — e.g. `project_visibility: public` with `use_community_health_files: false` — the migration now abstains instead of guessing, since `open_source` gates both LICENSE and community-health files together and there's no single boolean that can losslessly represent two previously-independent axes. Abstaining falls through to the question's own default, same as the pre-fix behavior, but now only for this narrow disagreement case instead of for every project.

Added 3 regression tests covering: community_health-only fallback, visibility+community_health agreement, and the disagreement/abstain case.
